### DISCUSSION TITLE
Add natural sorting function to cell content

### DISCUSSION THREAD

**Discussion by aborruso** [saulpw/visidata#1287] [Ideas]
Hi,
today I needed to sort (natural sorting) this list of values `15,1,2/AX,22,1/C,1/2,1/1,1/A,1/BA,2,3`, to obtain `1,1/1,1/2,1/A,1/BA,1/C,2,2/AX,3,15,22`.

I did this using the shell

```
echo "15,1,2/AX,22,1/C,1/2,1/1,1/A,1/BA,2,3" | tr , "\n" | sort -V | paste -sd, -
```

I'm not able to build a plugin, but I have build a Python hello world example, and some of the skilled users, could turn it into a plugin or a VisiData function:

```python
import re

_nsre = re.compile('([0-9]+)')
def natural_sort_key(s):
    return [int(text) if text.isdigit() else text.lower()
            for text in re.split(_nsre, s)]

mStr = '15,1,2/AX,22,1/C,1/2,1/1,1/A,1/BA,2,3'
list1 = mStr.split(',')

list1.sort(key=natural_sort_key)

joined_string = ",".join(list1)
```

The result is `1,1/1,1/2,1/A,1/BA,1/C,2,2/AX,3,15,22`.

The list separator should become an option.

**Comment by aborruso** [comment]
I'm adding two really interesting related URLs:

- https://www.gnu.org/software/coreutils/manual/coreutils.html#Version-sort-ordering (search "natural sort");
- a great python module and a cli to run natural sorting https://natsort.readthedocs.io/en/master/index.html

Using `natsort` cli, the "locale" option (a sort of human and non ASCII sorting):

```
echo "SNC-1-10-1r-1/A-2/10-10/1A2-10/A-1000/C1-a" | tr - "\n" | natsort -l  | paste -sd- -
```

gives me 

```
1-1/A-1r-2/10-10-10/1A2-10/A-1000/C1-a-SNC
```

**Comment by frosencrantz** [comment]
I had wanted to respond to this for some time.

It is pretty easy to add your function into your .visidatarc and then use
it with `addcol-expr` to create a new column with the sorted value, then type in `natsort(curcol)`  or `natsort(curcol, delimiter='/')`:

```
import re

_nsre = re.compile('([0-9]+)')
def natural_sort_key(s):
    return [int(text) if text.isdigit() else text.lower()
            for text in re.split(_nsre, s)]

def natsort(value, delimiter=','):
   list1 = value.split(delimiter)
   list1.sort(key=natural_sort_key)
   joined_string = delimiter.join(list1)
   return joined_string
```

I've wanted to do something similar with a column of values.  It looks like it is possible to create a custom column type, and then use https://github.com/SethMMorton/natsort/ to sort the values.  If the values were wrapped in a class, it looks like it would be possible to do something like this for adding natural sort order: https://github.com/SethMMorton/natsort/issues/60#issuecomment-397941351


Also, this seems like more a request for a new plugin from the community than from Saul or Anja directly.  I am wondering if we could add another issue tag for plugin requests or community volunteers.

**Comment by aborruso** [comment]
Thank you very much @frosencrantz

**Comment by saulpw** [comment]
Yeah, I agree with @frosencrantz, for the time being I think this should be more of a plugin, and we can see how it evolves.  Converting to a discussion, feel free to continue to toss around ideas.

**Comment by aborruso** [comment]
@frosencrantz @saulpw I have added it as "recipe", to my VisiData guide in Italian https://ondata.github.io/guidaVisiData/ricette/#applicare-il-natural-sorting-ai-contenuti-di-una-cella

Thank you again
